text/0000-glimmer-component-signature.md

```diff
@@ -0,0 +1,374 @@
+
+The goal of this change is to enable tooling to provide a more complete picture for developers of the ways in which Glimmer components can be used. The word "tooling" here is used in broad terms, including:
+ - API/design system documentation generators, like [Storybook] or [Hokulea]
+ - Language servers that provide hover information, go-to-definition, etc. like [TypeScript] or [uELS]
+ - Template-aware static analysis and typechecking systems, like [Glint] or [`els-addon-typed-templates`]
+
+[Storybook]: https://storybook.js.org/
+[Hokulea]: https://github.com/hokulea/hokulea
+[TypeScript]: https://www.typescriptlang.org/
+[uELS]: https://github.com/lifeart/ember-language-server
+[Glint]: https://github.com/typed-ember/glint
+[`els-addon-typed-templates`]: https://github.com/lifeart/els-addon-typed-templates
+
+## Detailed design
+
+Given a component with this template:
+
+```hbs
+<div class="notice notice-{{@kind}}" ...attributes>
+  {{yield this.dismiss}}
+</div>
+```
+
+This RFC proposes that a TypeScript backing class currently written like this:
+
+```ts
+import Component from '@glimmer/component';
+
+export interface NoticeArgs {
+  /** The kind of message displayed in this notice. Defaults to `info`. */
+  kind: 'error' | 'warning' | 'info' | 'success';
+}
+
+export default class Notice extends Component<NoticeArgs> {
+  // ...
+}
+```
+
+Would instead, if fully typed, be written like this:
+
+```ts
+import Component from '@glimmer/component';
+
+export interface GreetingSignature {
+  Element: HTMLDivElement;
+  Args: {
+    /** The kind of message displayed in this notice. Defaults to `info`. */
+    kind: 'error' | 'warning' | 'info' | 'success';
+  };
+  BlockParams: {
+    /**
+     * Any default block content will be displayed in the body of the notice.
+     * This block receives a single `dismiss` parameter, which is an action
+     * that can be used to dismiss the notice.
+     */
+    default: [dismiss: () => void];
+  };
+}
+
+export default class Greeting extends Component<GreetingSignature> {
+  // ...
+}
+```
+
+By nesting the existing `Args` type under a top-level _signature_ type, we create a place to provide additional type information for other aspects of a component's behavior. One key element of this design is that eases future evolution of the signature, as including additional keys with new meaning won't be a breaking change.
+
+In addition to `Args`, two other signature members are proposed here: `BlockParams` and `Element`. A detailed explanation of each of these follows.
+
+### `Args`
+
+The `Args` signature member works exactly as the top-level `Args` type parameter does in `@glimmer/component@1.x` prior to this RFC. That is, it determines the type of `this.args` in the component backing class, as well as acting as an anchor for human-readable documentation for specific arguments.
+
+A component that accepts no arguments can omit the `Args` key from its signature.
```

Comment:
This is a good addition but only appears for `Args` whereas actually applies to all sections. Should be mentioned for each section or _globally_ - I prefer for each section, then its better explainable.

Reply:
It's mentioned in each section.